### ISSUE TITLE
Enable running lints (that we run in CI) locally

### ISSUE THREAD

**Issue by janeyx99** [pytorch/pytorch#55847]
## 🚀 Feature
We currently run a bunch of lints on our codebase in CI to keep PyTorch clean (e.g., check out [this file](https://github.com/pytorch/pytorch/blob/master/.github/workflows/lint.yml)). It would be wonderful if we had an easy way to run those same lints locally in our own local PyTorch.

## Motivation
Imagine yourself as a contributor. You've just poured your heart out into the next big thing for PyTorch and you're super excited to see all the CI pass and be green for your new PR. You decide to take a well-deserved break as you wait for CI to all run, so you go do your favorite break activity and come back in 2 hours. You see a red X. Your heart skips a beat--what could have gone wrong?!?! 

Oh. It's just flake8. You quickly fix your badly indented code and repush, triggering another full run of CI before you finally land your change.

Now imagine how this situation would be different if you were able to quickly verify locally that your code wouldn't break any lints before you open the PR. Not only would you have landed your change faster with more satisfaction, you would have reduced the cost of an entire CI workflow. 

## Requirements + Their Difficulties
We would like the solution to...
1. Include as many quick and common lints. It is critical to include commonly broken lints while less critical to include lints that are less likely to be violated.
2. Verify that the versions of tools used in CI are the same as on a local server. It is possible for different versions of, say, `mypy` or `clang-tidy` to report different results.
     - This may be accomplished by accomplishing bullet 3. 
3. Verify that the code/lint commands in CI are semantically the same as the local version.
     - Some commands on CI are a lot more complicated than they have to be (see flake8). 
     - Some lints on CI require the codebase to be a certain way before being run. This may mess with local configurations.
4. Be platform-independent. This solution should work on most machines and possible local box configurations.
     - Some machines need proxy forwarding to download binaries.
     - Currently, all the code run in CI is fit for Ubuntu only.

## Possible (Barebone) Approaches

1. We update our Makefile to contain more targets that would each verify a part of our lint workflows on CI (e.g., one for `flake8`, one for `mypy`, `clang-tidy`...). We could have a target called `lint` that depends on all of them, so that one could run `make lint` and have all their code checked. 
2. We use a pre-commit hook that would run the code we have in our lint workflows on CI.
3. Feel free to suggest others

## Additional context

I believe the way to fully resolve this issue is to tackle it step by step. The preliminary method may not cover all the requirements just yet, but that's probably fine as long as there is progress.

**Comment by samestep** [comment]
To address the platform-dependence problem (4.), would we perhaps want to switch our Lint workflow to run as a matrix with [these virtual environments](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners#supported-runners-and-hardware-resources)?

- `ubuntu-18.04` (or `ubuntu-20.04` if we decide to upgrade)
- `macos-10.15`
- `windows-2019`

That would give us more assurance that the lints at least _run_ on those three platforms, although it wouldn't guarantee that they work on other Linux distros, and it also wouldn't guarantee that they're acting the same on each OS. Also, another issue with using a matrix is that it would effectively mean that three times as many lint jobs would be running on every PR, and I feel like that might be too much noise in the signals view.

I have an idea for a possible approach:

- Add specific linter targets and a general `lint` target to our `Makefile`, as @janeyx99 mentioned.
  - I don't know too much about using `make`, but perhaps we could either somehow run different lints in parallel or add some logic so that more expensive lints only run when necessary/requested? This might defeat the purpose or be too complicated, though.
  - In any case, I think it would be useful to standardize the way we use all the linters in `make lint`, so that they all store their output in a place that can later be accessed (see below).
- Condense `.github/workflows/lint.yml` into a _single_ `lint` job that just runs `make lint` (probably preceded by some setup to install the necessary software, although it would be really cool if we could include the installation steps in `make lint` itself).
  - The major downside of this consolidation is that it makes it harder to see at a glance (from the signals page on a PR) exactly which linter failed, without looking at the job logs. That might be reason enough not to go with this approach.
- Configure that single `lint` job to run on Ubuntu, macOS, and Windows via a matrix, so we would end up with only 3 lint jobs instead of a dozen or so.
- Add one more job in `.github/workflows/lint.yml` that takes the linter results from the three different OSes (using the stored output mentioned above) and fails if they don't all match each other.
  - Using [this technique](https://github.community/t/run-dependent-job-even-on-failure/115914), we should be able to have this aggregator job run even if the OS-specific `lint` jobs fail, so actual linter failures would show up on the OS-specific jobs rather than just on the aggregator job.
  - It would be nice if the aggregator job somehow let the @pytorch/pytorch-dev-infra team know when the linter outputs don't match up, since that should indicate something wrong with the Lint workflow rather than with the specific PR on which it's being run.

What do people think of this general idea? I kind of just threw this together so I don't think it's necessarily the right approach, but it might make a good starting point for discussing how to solve this issue.

**Comment by driazati** [comment]
A combination of the two seems best to me, where the `Makefile` has separate targets for each type of lint (e.g. `make flake8` and `make clang-tidy`) that we run on CI, and then we add another target for local testing 

```bash
make lint -j4  # run all lint jobs in parallel
```
that depends on all of those. This could dispatch out to a script or something that checks binary versions for `clang-*` to ensure that everything matches up. This way the core linting is the same on CI/local, but the invocation is different (so devs have 1 command but CI still shows each job separately).

**Comment by walterddr** [comment]
I guess these above mentioned approaches could be combined or separated. but a clear instruction section should be added to contributing.md. With @driazati 's point, this could produce a very easy-to-understand instruction
- run `make <specific-linter>` to lint something specific.
- run `make lint` before you create a PR.

From my point of view I would like to add one additional bulletin to the requirement section
5. Ability to run linter incrementally (provide fast iteration for developers)
  - let's say if I ran mypy and found a bug, fix it and rerun mypy again. currently it uses the preivous run cache to provide a lighting fast result. It would be nice if this could be apply to as much of the linters as possible.
  - This also mean if I rerun `make lint` multiple times it should be significantly faster.

**Comment by malfet** [comment]
We can also have `make quicklint`, which would run linter only on files from last commit (or files that have not been modified yet) and `make lint` which is slower

**Comment by samestep** [comment]
Btw regarding the Git pre-commit hook idea (2.), this blogpost discusses some important considerations: https://www.thoughtworks.com/insights/blog/pre-commit-don-t-git-hooked

**Comment by seemethere** [comment]
I think this could be closed by #56439? wdyt @driazati?

**Comment by zhouzhuojie** [comment]
I don't see docker mentioned here, can we just local run the same docker image the CI uses for lint? E.g. the `clang-tidy` image https://github.com/orgs/pytorch/packages/container/package/cilint-clang-tidy

**Comment by driazati** [comment]
docker is a no-go IMO since it isn’t supported on FB infra (dev servers) which a lot of PyTorch devs use

**Comment by janeyx99** [comment]
This has been done by lintrunner https://github.com/pytorch/pytorch/wiki/lintrunner
